miden-tx/src/host/mod.rs

```diff
@@ -38,19 +54,75 @@ impl<A: AdviceProvider> TransactionHost<A> {
-            account_delta: AccountDeltaTracker::new(&account),
+            state_tracker: StateTracker::new(&account),
             acct_procedure_index_map: proc_index_map,
+            notes: Vec::new(),
         }
     }
 
     /// Consumes `self` and returns the advice provider and account vault delta.
-    pub fn into_parts(self) -> (A, AccountDelta) {
-        (self.adv_provider, self.account_delta.into_delta())
+    pub fn into_parts(self) -> (A, AccountDelta, Vec<OutputNote>) {
+        (self.adv_provider, self.state_tracker.into_parts(), self.notes)
     }
 
     // EVENT HANDLERS
     // --------------------------------------------------------------------------------------------
 
+    fn on_note_created<S: ProcessState>(
+        &mut self,
+        process: &S,
+    ) -> Result<(), TransactionKernelError> {
+        let stack = process.get_stack_state();
+
+        // Stack:
+        // # => [aux, note_type, acct_id, tag, note_ptr, ASSET, RECIPIENT]
+        let aux = stack[0];
+        let note_type =
+            NoteType::try_from(stack[1]).map_err(TransactionKernelError::MalformedNoteType)?;
+        let sender =
+            AccountId::try_from(stack[2]).map_err(TransactionKernelError::MalformedAccountId)?;
+        let tag = NoteTag::try_from(stack[3])
+            .map_err(|_| TransactionKernelError::MalformedTag(stack[3]))?;
+        let asset = Asset::try_from([stack[8], stack[7], stack[6], stack[5]])
+            .map_err(TransactionKernelError::MalformedAsset)?;
+        let recipient = Digest::new([stack[12], stack[11], stack[10], stack[9]]);
+        let vault =
+            NoteAssets::new(vec![asset]).map_err(TransactionKernelError::MalformedNoteType)?;
+
+        let metadata = NoteMetadata::new(sender, note_type, tag, aux)
+            .map_err(TransactionKernelError::MalformedNoteMetadata)?;
+
+        let note = if metadata.note_type() == NoteType::Public {
+            let data = self.adv_provider.get_mapped_values(&recipient).ok_or(
+                TransactionKernelError::MissingNoteDetails(metadata, vault.clone(), recipient),
+            )?;
+            if data.len() != 12 {
+                return Err(TransactionKernelError::MalformedRecipientData(data.to_vec()));
+            }
+            let inputs_hash = Digest::new([data[0], data[1], data[2], data[3]]);
+            let script_root = Digest::new([data[4], data[5], data[6], data[7]]);
+            let serial_num = [data[8], data[9], data[10], data[11]];
+            let input_els = self.adv_provider.get_mapped_values(&inputs_hash);
+            let script_data = self.adv_provider.get_mapped_values(&script_root).unwrap_or(&[]);
+
+            let inputs = NoteInputs::new(input_els.map(|e| e.to_vec()).unwrap_or_default())
+                .map_err(TransactionKernelError::MalformedNoteInputs)?;
+            let script = NoteScript::try_from(script_data)
+                .map_err(|_| TransactionKernelError::MalformedNoteScript(script_data.to_vec()))?;
```

Comment:
This is fine for now, but in the future, we'll also need to validate that the script data is consistent with script hash. This will become simpler once we move to MAST-based script serialization (i.e., we will not need to compile the script). Let's add a `TODO` about it here and also create an issue to come back to this later.